File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque


def solution(land):
    n = len(land)
    m = len(land[0])
    
    visited = [[False for _ in range(m)] for _ in range(n)]
    
    answer = 0
    
    def bfs(r,c, type):
        que = deque([[r,c]])
        visited[r][c]=True
        size = 1
        land[r][c]=type
        while que:
            r,c = que.popleft()
            for nr, nc in [[r-1, c],[r,c-1],[r+1,c],[r,c+1]]:
                ## 범위에 나감 // 이미 방문함
                if not(0<=nr<n) or not(0<=nc<m) or visited[nr][nc]:
                    continue
                if land[nr][nc]==1:
                    visited[nr][nc]=True
                    land[nr][nc]=type
                    size+=1
                    que.append([nr, nc])
        return size
        
    type = 2
    sizeDic={}
    for i in range(n):
        for j in range(m):
            if not(visited[i][j]) and land[i][j]==1:
                sizeDic[type]=bfs(i,j, type)
                type+=1
                

    def getSize(lst):
        size =0
        for k in lst:
            size+=sizeDic[k]
        return size

    for j in range(m):
        temp = []
        for i in range(n):
            if land[i][j]!=0 and land[i][j] not in temp:
                temp.append(land[i][j])
        answer = max(answer, getSize(temp))



            
        
    return answer

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
function solution(land) {
    const [N,M] = [land.length, land[0].length]
    const visited=Array.from({length:N},()=>Array.from({length:M}).fill(false))
    
    const directions =[
        [1,0],[-1,0],[0,1],[0,-1]
    ]
    const chuncks = Array.from({length:M}).fill(0)
    
    function bfs(i,j){
        const que = [[i,j]]
        visited[i][j]=true
        let [size, min_c, max_c] = [1,j,j]
        while(que.length){
            const [r,c] = que.shift()
            for(let [dr, dc] of directions){
                const [nr, nc] = [r+dr, c+dc]
                if(0>nr || N<=nr || 0>nc || M<=nc||land[nr][nc]==0||visited[nr][nc]){continue}
                visited[nr][nc]=true
                size+=1
                que.push([nr, nc])
                min_c = Math.min(min_c, nc)
                max_c = Math.max(max_c, nc)
                
            }
        }
        // chuncks.push({size,min_c, max_c})
        for(let c=min_c;c<=max_c;c++){
           chuncks[c]+=size 
        }
        // console.log(i,j,size, min_c, max_c)
    }
    
    for(let i=0;i<N;i++){
        for(let j=0;j<M;j++){
            if(land[i][j]==1 && !visited[i][j]){
                bfs(i,j)
            }
        }
    }
    // console.log(chuncks)
    const max_size = Math.max(...chuncks)
  
    return max_size;
}